scripts/trim_indel_ends.py

```diff
@@ -0,0 +1,402 @@
+"""
+    File name: trim_indel_ends.py
+    Author: Yutong Qiu
+    Date: Nov 20
+    Description: For each pair of alinment between reads, if the best alignment includes indels at the beginning and ends
+        of the reads, then trim off the indels.
+    Input: 
+        - old sets of reads for each sample
+        - aln_stats
+    Output: 
+        - new set of reads for each sample.
+
+    Trimming logic alternatives:
+        - Trim prefix and suffix regardless of the rest of the alignment --- set a threshold
+        - Trim prefix and suffix only when the rest of the alignment is all matches --- set a threshold
+"""
+
+import pandas as pd
+from collections import defaultdict, Counter
+import re
+import edlib
+from Bio import SeqIO
+from Bio.SeqRecord import SeqRecord
+from Bio.Seq import Seq
+import matplotlib.pyplot as plt
+from tqdm import tqdm
+import os
+import sys
+import argparse
+import datetime
+
+#--------------------------------------------------------
+# parser
+#--------------------------------------------------------
+# positional/optional arguments
+posList = ['manifest','unique_isoform_aln']
+optList = ['mode', 'threshold']
+sys.stdout.write('\n'); sys.stdout.flush()
+
+# author and version info
+usage = sys.argv[0] + ' <options> ' + ' '.join(posList) + '\n'
+author = 'Author: Yutong Qiu (yutongq@andrew.cmu.edu)'
+version = 'Version: 1.0.0.0 (2022-11-23)'
+description = '\nDescription: This program trims isoform sequences to reduce false positives. Output trimmed sequences in trimmed.fasta'
+
+parser = argparse.ArgumentParser(usage = '\n'.join([usage, author, version, description]), formatter_class=argparse.RawTextHelpFormatter)
+# positional arguments
+parser.add_argument(posList[0], type=str, help='string\tinput manifest csv file in format: sample,fa.  e.g. /in/manifest.csv')
+parser.add_argument(posList[1], type=str, help='string\talignment between similar isoforms within each window.  e.g. /in/unique_isoforms_and_aln_stats.tsv')
+# optional arguments
+parser.add_argument('-m', '--'+optList[0], type=str, metavar='', default='lenient', help='string\tMode of trimming. "Strict" only trims when sequences match exactly after trimming; "Lenient" trims whenever prefix and suffix indel numbers below threshold, defauls to "Lenient"')
+parser.add_argument('-t', '--'+optList[1], type=int, metavar='', default=10000, help='int\tThreshold for trimming, defaults to 10000 (trim every prefix and suffix)')
+
+# strict positional/optional arguments checking
+argsDict = vars(parser.parse_args())
+sys.stdout.write('\n' + datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S') + ' - ' + sys.argv[0] + ' - Parsing Input Arguements...\n\n'); sys.stdout.flush()
```

Comment:
This is a suggestion, not a recommendation for a change -- it is something I can do when I finally get around to moving scripts from `scripts` into `src`.

That said, I would suggest that every `print` statement be given to a logger instance rather than using `print` or `sys.stdout`, etc.

[Here is an example of how to set up a logger](https://github.com/collaborativebioinformatics/isocomp/blob/ea01bdd9579330ca79ae521f8e8b66ecb41dd249/src/isocomp/create_windows.py#L24)

Then in your script, you can send things to the logger like this:

```logger.debug(sys.argv)```

or something like that. That way you can test the logger output -- [example](https://github.com/cmatKhan/lmdemo/blob/main/tests/test_lmdemo.py) -- and the `print` etc can be more easily configured in a notebook if a user is using these functions as an API, or with settings in argparse if the user is calling a cmd line function. 

This comment applies to all `stdout` and `print` statements

Reply:
This is an example of another software package which uses logging and offers a tutorial on how to set up the logger to a user who is using the software in a jupyter notebook:

https://github.com/abyzovlab/CNVpytor/blob/master/examples/PythonLibraryGuide.ipynb

and this is an example of setting the logger on the cmd line

[argparse setup](https://github.com/cmatKhan/pycallingcards/blob/c09ed401379ba5c60375e66db1a7db0cbab9369e/pycallingcards/__main__.py#L33)

[usage in main](https://github.com/cmatKhan/pycallingcards/blob/c09ed401379ba5c60375e66db1a7db0cbab9369e/pycallingcards/__main__.py#L133)

Reply:
That looks good. I am not very familiar with the convention for logging. Maybe we could discuss that in a meeting and set up a convention for this repository.